packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +462,339 @@ class ShellRoute extends RouteBase {
+/// [StatefulShellRoute] (using the builder function) as input. In other words,
+/// builder and pageBuilder must both be provided when using a custom page for
+/// a StatefulShellRoute. For example:
+///
+/// ```
+/// final GoRouter _router = GoRouter(
+///   initialLocation: '/a',
+///   routes: <RouteBase>[
+///     StatefulShellRoute(
+///       builder: (BuildContext context, GoRouterState state,
+///           Widget navigationContainer) {
+///         return ScaffoldWithNavBar(body: navigationContainer);
+///       },
+///       pageBuilder:
+///           (BuildContext context, GoRouterState state, Widget statefulShell) {
+///         return NoTransitionPage<dynamic>(child: statefulShell);
+///       },
+///       branches: <StatefulShellBranch>[
+///         /// The first branch, i.e. root of tab 'A'
+///         StatefulShellBranch(routes: <RouteBase>[
+///           GoRoute(
+///             parentNavigatorKey: _tabANavigatorKey,
+///             path: '/a',
+///             builder: (BuildContext context, GoRouterState state) =>
+///               const RootScreen(label: 'A', detailsPath: '/a/details'),
+///           ),
+///         ]),
+///         /// The second branch, i.e. root of tab 'B'
+///         StatefulShellBranch(routes: <RouteBase>[
+///           GoRoute(
+///             parentNavigatorKey: _tabBNavigatorKey,
+///             path: '/b',
+///             builder: (BuildContext context, GoRouterState state) =>
+///               const RootScreen(label: 'B', detailsPath: '/b/details/1'),
+///           ),
+///         ]),
+///       ],
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// To access the current state of this route, to for instance access the
+/// index of the current route branch - use the method
+/// [StatefulShellRouteState.of]. For example:
+///
+/// ```
+/// final StatefulShellRouteState shellState = StatefulShellRouteState.of(context);
+/// ```
+///
+/// See [Stateful Nested Navigation](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_nested_navigation.dart)
+/// for a complete runnable example using StatefulShellRoute.
+class StatefulShellRoute extends ShellRouteBase {
+  /// Constructs a [StatefulShellRoute] from a list of [StatefulShellBranch]es,
+  /// each representing a separate nested navigation tree (branch).
   ///
-  /// Similar to GoRoute builder, but with an additional child parameter. This
-  /// child parameter is the Widget built by calling the matching sub-route's
-  /// builder.
-  final ShellRouteBuilder? builder;
+  /// A separate [Navigator] will be created for each of the branches, using
+  /// the navigator key specified in [StatefulShellBranch]. Note that unlike
+  /// [ShellRoute], a builder must always be provided when creating a
+  /// StatefulShellRoute. The pageBuilder however is optional, and is used
+  /// in addition to the builder.
+  StatefulShellRoute({
+    required this.branches,
+    required super.builder,
```

Comment:
You can tight the route type to be safer
```suggestion
    required ShellRouteBuilder super.builder,
```

Reply:
Yes, I wasn't too happy with how builder and pageBuilder worked compared with other routes. However, there is the dilemma how to handle the builders since they are a bit more complex than `ShellRoute`, due to the fact that you have to build a shell wrapping a custom shell body. But I think the solution in the other branch, using a `StatefulShellBuilder`, is possibly a way to go, since it again involves the use of builder **_or_** pageBuilder.  

```dart
builder: (StatefulShellBuilder shellBuilder) {
  return shellBuilder.buildShell((BuildContext context,
          GoRouterState state, ShellNavigatorContainer child) =>
      ScaffoldWithNavBar(body: child));
}
// Or:
pageBuilder: (StatefulShellBuilder shellBuilder) {
  final Widget shell = shellBuilder.buildShell((BuildContext context, 
          GoRouterState state, ShellNavigatorContainer child) =>
      ScaffoldWithNavBar(body: child));
  return NoTransitionPage<dynamic>(child: shell);
}
```